dev-docs/design/how-lit-works.md

```diff
@@ -0,0 +1,302 @@
+// the count for the same template strings array even though the dynamic values are different.
+let f = () => fn('abc');
+let g = () => fn('def');
+
+f();
+g();
+```
+
+Prints:
+
+```bash
+literal 'static portion of template with a dynamic value:' seen 1 time(s)
+literal 'static portion of template with a dynamic value:' seen 2 time(s)
+```
+
+lit-html uses the template strings array argument as a cache key to only do template preparation work on the first render to the page.
+
+## The HTML `<template>` Element
+
+The [`<template>` element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/template) is a container for _inert_ HTML content. Its contents are parsed, but not rendered: scripts don't run, styles don't apply, Custom Elements don't upgrade, etc.
+
+During parsing, or when setting `innerHTML`, nodes that would normally become children in other elements, are moved into the template's _content_ fragment: a `DocumentFragment` accessible via the `.content` property. A template's contents can be cloned or imported into a Document, which makes the cloned nodes active.
+
+## Parts
+
+A `Part` is a lit-html concept and represents the location of an expression in the `html` tagged template literal:
+
+| Part                   | Description                                                        | Authored                               |
+| ---------------------- | ------------------------------------------------------------------ | -------------------------------------- |
+| `ChildPart`            | Expressions in HTML child position                                 | `` html`<div>${...}</div>`  ``         |
+| `AttributePart`        | Expressions in HTML attribute value position                       | `` html`<input id="${...}">`  ``       |
+| `BooleanAttributePart` | Expressions in a boolean attribute value (name prefixed with `?`)  | `` html`<input ?checked="${...}">`  `` |
+| `EventPart`            | Expressions in an event listener position (name prefixed with `@`) | `` html`<input @click=${...}>`  ``     |
+| `PropertyPart`         | Expressions in property value position (name prefixed with `.`)    | `` html`<input .value=${...}>`  ``     |
+| `ElementPart`          | Expressions on the element tag                                     | `` html`<input ${...}>`  ``            |
+
+In all the cases above the authored code pass an expression into `${<expr>}` which represents a dynamic binding to the template, and the different part types implement how the value is committed to the DOM. For instance the `EventPart` in `` html`<input @click=${() => console.log('clicked')}`  `` will take the user provided function, and manage `addEventListener` and `removeEventListener` calls on the DOM such that the passed function is called when the click event is triggered.
+
+Knowing about Parts is useful when [writing custom directives](https://lit.dev/docs/templates/custom-directives/#parts).
+
+# Rendering
```

Comment:
This is the meat of the doc and I want a TLDR up front with 1 sentence each on: define, prepare, create, update.

1. Define: ``` const template = value => html`<div>${value}</div>` ```
2. Render: `render(template('hi'), element)`
  A. Prepare: The first time render is called for each unique template, create a `<template>` element containing the static html and  Lit `Parts` to update the dynamic content.
  B. Create: The first time render is called for a given node and template, clone the template into the DOM and set all the dynamic content.
  C. Update: Whenever render is called again with the same template, update only the dynamic parts fhat have changed.

Reply:
Done